New sample: Show location history (#556)
Esri/arcgis-runtime-samples-java: map_view/show-location-history/README.md

# Show location history

Display your location history on the map.

![Image of show location history](ShowLocationHistory.png)

## Use case

You can track device location history and display it as lines and points on the map. The history can be used to visualize how the user moved through the world, to retrace their steps, or to create new feature geometry. For example, an unmapped trail could be added to the map using this technique.

## How to use the sample

> Click the button to start tracking your location, which will appear as points on the map. A line will connect the points for easier visualization. Click the button again to stop updating the location history. This sample uses a simulated data source to allow the sample to be useful on devices without a location data source.

Reviewer: I think we need to specify/reword this slightly to make clear what happens when the app opens: when it opened I assumed the moving point _was_ tracking location (a moving point on a map) before realising that tracking makes an actual path. 

I've suggested removing the reference to "allow the sample to be useful on devices without a location data source" since that instruction is repeated in the "About the data" section.

```suggestion
The sample loads with a moving simulated location data source. Click the button to start tracking the location, which will appear as red points on the map. A green line will connect the points for easier visualization. Click the button again to stop updating the location history.
```

> ## How it works

Reviewer: Kinda feel we might be a bit thin on the ground here instruction wise. E.g. we don't have a mention of getting the `LocationDisplay` from the `MapView`. Is there another sample in the pipeline that will be going over the basics of location displays? If so, we don't need to be too explicit about that here but I think it's worth dropping in mention of `locationDisplay` here since it's referred to in the Relevant API section. 

A good way to think about it: If it's not relevant enough to be in the "How it works", it's not relevant enough to go in the Relevant API section.

Reviewer: Also the autopan is an interesting feature of the sample, and it's referenced in the Relevant API: some reference to that instruction here would be good to do also.

Author: I'm not mentioning `SimulationParameters` in the "How it works" section. Would it be ok to also remove it from the Relevant API section?

1. Create a `GraphicsOverlay` to show each point and another `GraphicsOverlay` for displaying the route line.

> 2. Create a `SimulatedLocationDataSource` and initialize it with a polyline. Start the `SimulatedLocationDataSource` to begin receiving location updates.

Reviewer: ```suggestion
2. Create a `SimulatedLocationDataSource` and initialize it with a `Polyline`. Start the `SimulatedLocationDataSource` to begin receiving location updates.
```

3. Use a `LocationChangedListener` on the `simulatedLocationDataSource` to get location updates.
4. When the location updates store that location, display a point on the map at the location, and re-create the route polyline.

## Relevant API

* LocationDataSource
* LocationDataSource.Location
* LocationDataSource.LocationChangedEvent
* LocationDataSource.LocationChangedListener

> * LocationDataSource.Position

Reviewer: This has always been a slightly contentious deviation by Java from the rest of the samples team to not include inner classes/methods in the "Relevant API" section. The idea being that `LocationDataSource` in itself is clear enough for this section, and that the inner workings of it are described within the "How it Works" section and code well. So from that perspective, I'd be in favour of removing these and keeping only `LocationDataSource`.  It's not a biggy if you want to keep them in though.

Reviewer: Also: I guess we're not really using `LocationDataSource` in this sample, but `SimulatedLocationDataSource`? The only place we use `LocationDataSource` is to create a listener, and there's no reference to the class in "How it works": only in the "About the Data" section which is what got me thinking.

In which case do we only want to include `LocationDataSource.LocationChangedEvent` and `LocationDataSource.LocationChangedListener`

* LocationDisplay.AutoPanMode

> * MapView.LocationDisplay

Reviewer: e.g. this I am in favour of keeping because it specifically links the MapView to a LocationDisplay: if we just had `MapView` it wouldn't be clear its link to the sample other than the usual "display a map on the map view".

* SimulatedLocationDataSource
* SimulationParameters

## About the data

> A custom set of points is used to create a `Polyline` and initialize a `SimulatedLocationDataSource`. The simulated location data source enables easier testing and allows the sample to be used on devices without an actively updating GPS signal. To track a user's real position, use the `LocationDataSource` instead. 

Reviewer: Is there a sample coming through for `LocationDataSource`? If so, we should come back to this readme when it's completed and hyperlink to it here. Or alternatively I might like to see some API Ref when it's available to hyperlink this class to for users wanting to easily access more info (not available until 100.10.0 release)
```suggestion
A custom set of points (provided in JSON format) is used to create a `Polyline` and initialize a `SimulatedLocationDataSource`. The simulated location data source enables easier testing and allows the sample to be used on devices without an actively updating GPS signal. To track a user's real position, use `LocationDataSource` instead. 
```

# Show location history

Display your location history on the map.

> ![](ShowLocationHistory.png)

Reviewer 2: Don't forget to add a screenshot (and maybe even a fancy gif for the SV?)

Reviewer 2: Sorry, what I meant here was the 'alt' text is missing. It should say:
`![Image of show location history](ShowLocationHistory.png)`

## Use case

You can track device location history and display it as lines and points on the map. The history can be used to visualize how the user moved through the world, to retrace their steps, or to create new feature geometry. An unmapped trail, for example, could be added to the map using this technique.

## How to use the sample

> Click the button to start tracking your location, which will appear as points on the map. A line will connect the points for easier visualization. Click the button again to stop updating the location history.

Reviewer 2: I might suggest adding a mention of the simulated location data source here already, since this section by itself would give the users the impression they are seeing some sort of 'live' location?

Author: Perhaps I could also add a line like: "To track a user's real position, use the `LocationDataSource` instead"?

Reviewer 2: Yes, that's a good call. It's important to make clear to the users that they are not actually driving through LA when they run this sample ;)

## How it works

> 1. If necessary, request location permission from the operating system.

Reviewer 2: We don't have this in Java, so I think we can leave this out. A potential sample with the NMEA api would handle passing in 'real' data into `LocationDataSource`

2. Create a `GraphicsOverlay` to show each point and another `GraphicsOverlay` for displaying the route line.
3. Create a `SimulatedLocationDataSource` and initialize it with a polyline. Start the `SimulatedLocationDataSource` to begin receiving location updates.
4. Use a `LocationChangedListener` on the `simulatedLocationDataSource` to get location updates.
5. When the location updates store that location, display a point on the map at the location, and re-create the route polyline.

## Relevant API

* LocationDataSource.Position

> * LocationDataSource

Reviewer 2: We could put this one at the top of the list, so they are sorted 'alphabetically'.
Not that there's a general tendenct only to put top-level classes in here, but since the inner classes `Location`, `LocaitonChangedEvent`, etc. are all quite important, this is ok.

* LocationDataSource.Location
* LocationDataSource.LocationChangedEvent
* LocationDataSource.LocationChangedListener
* LocationDisplay.AutoPanMode
* MapView.LocationDisplay
* SimulatedLocationDataSource
* SimulationParameters

## About the data

> The sample uses a dark gray basemap with lime and red graphics. 

Reviewer 2: I don't think we need this description, it's quite obvious and the type of basemap/graphics aren't really that relevant to the way the API works

Display your location history on the map.

![Image of show location history](ShowLocationHistory.gif)

## Use case

You can track device location history and display it as lines and points on the map. The history can be used to visualize how the user moved through the world, to retrace their steps, or to create new feature geometry. For example, an unmapped trail could be added to the map using this technique.

## How to use the sample

The sample loads with a moving simulated location data source. Click the button to start tracking the location, which will appear as red points on the map. A green line will connect the points for easier visualization. Click the button again to stop updating the location history.
## How it works

1. Create a `GraphicsOverlay` to show each point and another `GraphicsOverlay` for displaying the route line.

> 2. Create a `SimulatedLocationDataSource` and initialize it with a `Polyline`. Start the `SimulatedLocationDataSource` to begin receiving location updates.

Reviewer: let's add in `SimulationParameters` here as you were asking about: we can also be a bit clearer about how the `Polyline` and parameters are being added, since technically, we aren't initializing the `SimulatedLocationDataSource` with them: we are creating a new one and then setting its location. 
```suggestion
2. Create a `SimulatedLocationDataSource` and call its `setLocations()` method, passing the route `Polyline` and new `SimulatedParameters` as parameters. Start the `SimulatedLocationDataSource` to begin receiving location updates.
```

3. Use a `LocationChangedListener` on the `simulatedLocationDataSource` to get location updates.
4. When the location updates store that location, display a point on the map at the location, and re-create the route polyline.

## Relevant API

* LocationDataSource.LocationChangedEvent
* LocationDataSource.LocationChangedListener
* LocationDisplay.AutoPanMode
* MapView.LocationDisplay
* SimulatedLocationDataSource
* SimulationParameters

## About the data

> A custom set of points (provided in JSON format) is used to create a `Polyline` and initialize a `SimulatedLocationDataSource`. The simulated location data source enables easier testing and allows the sample to be used on devices without an actively updating GPS signal. To track a user's real position, use `NMEALocationDataSource` instead. 

Reviewer: since it's not strictly used in an initializer, slight word tweak here: let me know if you think it doesn't make sense, we can try to find a better word! 
```suggestion
A custom set of points (provided in JSON format) is used to create a `Polyline` and configure a `SimulatedLocationDataSource`. The simulated location data source enables easier testing and allows the sample to be used on devices without an actively updating GPS signal. To track a user's real position, use `NMEALocationDataSource` instead. 
```

## Use case

You can track device location history and display it as lines and points on the map. The history can be used to visualize how the user moved through the world, to retrace their steps, or to create new feature geometry. For example, an unmapped trail could be added to the map using this technique.

## How to use the sample

> Click the button to start tracking your location, which will appear as points on the map. A line will connect the points for easier visualization. Click the button again to stop updating the location history. This sample uses a simulated data source to allow the sample to be useful on desktop/non-mobile devices.

Reviewer 2: I think it'd be more percise to say `This sample uses a simulated data source to allow the sample to be useful on devices without a location data source`. This is since just because a device is desktop, doesn't mean it won't have the means of getting it's location. For example most desktop devices can get a rough location via WiFi, etc.

Author: Sounds good! I could also add this to the use case: `You can use an existing dataset in place of a connection to a GPS device. A simulated data source is useful if you're giving a demonstration or are working with previously collected data.` from [Setting up the simulation](https://desktop.arcgis.com/en/arcmap/10.3/map/display-gps-data/setting-up-the-simulation.html). But that could be redundant!
